examples/plot_connectivity.py

```diff
@@ -0,0 +1,115 @@
+"""
+=======================
+07. Plot Connectivity
+=======================
+
+This example demonstrates how to modify the network the network connectivity.
+"""
+
+# Author: Nick Tolley <nick nicholas_tolley@brown.edu>
+
+import os.path as op
+
+###############################################################################
+# Let us import ``hnn_core``.
+
+import hnn_core
+from hnn_core import read_params, Network, simulate_dipole
+
+hnn_core_root = op.dirname(hnn_core.__file__)
+
+###############################################################################
+# Then we read the parameters file
+params_fname = op.join(hnn_core_root, 'param', 'default.json')
+params = read_params(params_fname)
+
+###############################################################################
+# To explore how to modify network connectivity, we will start with simulating
+# the evoked response from the
+# :ref:`evoked example <sphx_glr_auto_examples_plot_simulate_evoked.py>`, and
+# explore how it changes with new connections. We first instantiate build
+# the network.
+net = Network(params)
+
+###############################################################################
+# Instantiating the network comes with a predefined set of connections that
+# reflect the canonical neocortical microcircuit. ``net.connectivity_list``
+# is a list of dictionaries which detail every cell-cell, and drive-cell
+# connection.
+print(len(net.connectivity_list))
+print(net.connectivity_list[0])
+
+###############################################################################
+# Next we add the distal drive. Note that adding drives creates new drive-cell
+# connections that are appended to ``net.connectivity_list``
+weights_ampa_d1 = {'L2_basket': 0.006562, 'L2_pyramidal': .000007,
+                   'L5_pyramidal': 0.142300}
+weights_nmda_d1 = {'L2_basket': 0.019482, 'L2_pyramidal': 0.004317,
+                   'L5_pyramidal': 0.080074}
+synaptic_delays_d1 = {'L2_basket': 0.1, 'L2_pyramidal': 0.1,
+                      'L5_pyramidal': 0.1}
+net.add_evoked_drive(
+    'evdist1', mu=63.53, sigma=3.85, numspikes=1, weights_ampa=weights_ampa_d1,
+    weights_nmda=weights_nmda_d1, location='distal',
+    synaptic_delays=synaptic_delays_d1, seedcore=4)
```

Comment:
I realize there is a lot of boilerplate here that distracts you from the main purpose of the example. We might want to brainstorm some ideas with @cjayb how to reduce this. Is it acceptable to use `add_drives_from_params` or is the plan to deprecate it?

Reply:
I totally agree: there's a lot of boilerplate here that distracts from the main message. Maybe we do want to keep `add_drives_from_params` in the API? One major concern I have is that the current parameter file structure is a mess and error-prone (the std is the fourth element of a list in a dict, the keys of which have to be parsed to figure out where we are...).

Maybe the solution isn't to deprecate the option, but fix the parameter file definition? I would be OK with using `add_drives_from_params` in this example for now (your call @ntolley), and deferring the decision on what to do about params for later.

Reply:
Isn't there another param called `legacy_mode` ? I didn't understand what that does. If I set it to False will the default drives still be added?

Reply:
It's related. I had to include it for two reasons: i) param-files allow all kinds of rubbish values for parameters that don't end up mattering (e.g. when `T > tstop`), and ii) HNN GUI creates a drive cell for each real cell class (270 in total), _regardless of whether or not the drive has AMPA/NMDA innervations to those cells_.

The only place `legacy_mode=False` is currently in `test_drives.py`, where we do sanity checking of inputs. Note that a `net` created with this "mode" won't run! The reason is that `_connect_celltypes` has (had?) [hard-coded assumptions](https://github.com/jonescompneurolab/hnn-core/blob/33d70443301dcd382d9c775195fd9adc4eae7f33/hnn_core/network_builder.py#L465) about the order of cells and drives in the gid list (hence 'legacy'). I'm actually waiting to see whether this PR could already allow us to do away with `legacy_mode`. If not, I'll certainly do so soon!

Reply:
@ntolley to be specific: What I think _should_ work, but currently doesn't, is this:

```python
    net = Network(params, legacy_mode=False)
    net.add_evoked_drive('early_distal', mu=10, sigma=1, numspikes=1,
                         location='distal', weights_ampa={
                             'L2_pyramidal': 1e-4,
                             'L5_pyramidal': 5e-5
                         })
    dpl = simulate_dipole(net, n_trials=1)
```

where

```python
print(net.gid_ranges)
{'L2_basket': range(0, 35), 'L2_pyramidal': range(35, 135), 'L5_basket': range(135, 170), 'L5_pyramidal': range(170, 270), 'early_distal': range(270, 470)}
```

Note the last range is from 270 to 470 (200 pyramidal cells targeted). When `legacy_mode=True`, this range goes to 540, because even though basket cells are not targeted, they are included in the GID ranges. The problem lies in `_connect_celltypes` and now also in your new `_all_to_all_connect`, I think.

I don't mind letting this PR go first, and then tackling the issue separately. I just wanted to let you know what it's about. Maybe you stumble on an easy fix.

Reply:
Let's move this to a separate issue?

Btw, @ntolley what do you think of setting `add_drives_from_params=True` and reducing the code to focus on connectivity. You can cross-link to the evoked example telling users how they can set drives ...

Reply:
That sounds pretty reasonable to me! Also this will free up some space for plotting/visualizing connectivity (a desperately needed followup to this PR).